plugins/single_plugins/scale.cpp

```diff
@@ -0,0 +1,1277 @@
+            fade_in(view);
+            output->focus_view(view, true);
+
+            return;
+        }
+
+        if (!scale_data.count(view))
+        {
+            return;
+        }
+
+        int row = scale_data[view].row;
+        int col = scale_data[view].col;
+
+        if ((state == WLR_KEY_RELEASED) &&
+            ((key == KEY_ENTER) || (key == KEY_ESC)))
+        {
+            input_release_impending = false;
+        }
+
+        if ((state != WLR_KEY_PRESSED) ||
+            wf::get_core().get_keyboard_modifiers())
+        {
+            return;
+        }
+
+        switch (key)
+        {
+          case KEY_UP:
+            row--;
+            break;
+
+          case KEY_DOWN:
+            row++;
+            break;
+
+          case KEY_LEFT:
+            col--;
+            break;
+
+          case KEY_RIGHT:
+            col++;
+            break;
+
+          case KEY_ENTER:
+            input_release_impending = true;
+            deactivate();
+            select_view(current_focus_view);
+
+            return;
+
+          case KEY_ESC:
+            input_release_impending = true;
+            deactivate();
+            output->focus_view(initial_focus_view, true);
+            initial_focus_view = nullptr;
+            output->workspace->request_workspace(initial_workspace);
+
+            return;
+
+          default:
+            return;
+        }
+
+        if ((grid_rows > 1) && (grid_cols > 1) &&
+            (grid_last_row_cols > 1))
+        {
+            /* when moving to and from the last row, the number of columns
+             * may be different, so this bit figures out which view we
+             * should switch focus to */
+            if (((key == KEY_DOWN) && (row == grid_rows - 1)) ||
+                ((key == KEY_UP) && (row == -1)))
+            {
+                auto p = col / (float)(grid_cols - 1);
+                col = p * (grid_last_row_cols - 1);
+                col = std::clamp(col, 0, grid_last_row_cols - 1);
+            } else if (((key == KEY_UP) && (row == grid_rows - 2)) ||
+                       ((key == KEY_DOWN) && (row == grid_rows)))
+            {
+                auto p = (col + 0.5) / (float)grid_last_row_cols;
+                col = p * grid_cols;
+                col = std::clamp(col, 0, grid_cols - 1);
+            }
+        }
+
+        if (row < 0)
+        {
+            row = grid_rows - 1;
+        }
+
+        if (row >= grid_rows)
+        {
+            row = 0;
+        }
+
+        int current_row_cols = (row == grid_rows - 1) ?
+            grid_last_row_cols : grid_cols;
+        if (col < 0)
+        {
+            col = current_row_cols - 1;
+        }
+
+        if (col >= current_row_cols)
+        {
+            col = 0;
+        }
+
+        view = find_view_in_grid(row, col);
+        if (view && (current_focus_view != view))
+        {
+            // view_focused handler will update the view state
+            output->focus_view(view, true);
```

Comment:
Here, do we really want to change the stacking order? (which is what the second parameter indicates) I'd think we want to not change the stacking order except when exiting (ENTER or ESC)

Reply:
Make a PR?